### PR TITLE
Run prettier and pre-commit hooks only on staged files

### PR DESCRIPTION
**Pull request by vjousse**
## :wrench: Problem

Currently, the pre-commit hook managed by husky run prettier on all the files for each commit, even if the files were not changed. It slows down the dev process unnecessarily most of the time.

## :cake: Solution

Run linting only on staged files when commiting.

## :rotating_light:  Points to watch / comments

My first attempt was to keep using husky with https://github.com/lint-staged/lint-staged. It works well for prettier but doesn't work for ruff as it doesn't manage `pipenv run` nor `virtualenv` correctly. It looks very "javascript centric".

My second and final attempt was to use https://pre-commit.com/. It's written in Python, it can manage every language (with hooks or using shell commands) and is well suited for our environment as we already use `pipenv` and Python.

`npm` commands removed:
- `format:check` => replaced by `lint:all` and changed in the CI file

`npm` commands added:
- `lint:prettier` lint one file using prettier. Used by other npm commands.
- `lint:prettier:all` lint all files in the project using prettier.
- `lint:ruff:all` lint all python files using ruff check and format (with sort option selected).
- `lint:all` prettier and ruff lint on all files.
- `fix:prettier` lint and fix the errors using prettier for one file (used in pre-commit hook)
- `fix:prettier:all` lint and fix the errors using prettier for all the files.
- `fix:ruff:all` lint all python files using ruff check and format (with sort option selected) and fix the errors
- `fix:all` prettier and ruff fix on all files

A `pyproject.toml` file was added to tell ruff to not check the `data/` directory.

I've added some simple checks with pre-commit like yaml format, line-ending checks (avoid mixing windows/linux/macox line ending), trailing-spaces removal and a check to prevent commiting directly on master (I suppose the git repo is already configured to avoid that, but I like to have it locally before I push).

## :desert_island: How to test

First, uninstall husky:

    npx husky uninstall

Check that the command `git config --get core.hookspath` returns nothing. If it gives you something like `.husky/\_` unset it manually using `git config --unset-all core.hooksPath`

Install pre-commit:

    pipenv install
    pipenv run pre-commit install

Test that it's working as expected by issuing the following command:

    pipenv run pre-commit run --all-files

You should get something like that:
image

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[project]
+requires-python = ">= 3.11"
+
+[tool.ruff]
+exclude = ["data/"]
```
Comment (n1k0) [inline pyproject.toml:5]: Lots of Python code in there, why would we want to skip formatting it?

File: data/bwapi/README.txt
Comment (n1k0) [inline data/bwapi/README.txt:1]: this one too

File: data/Makefile [hunk 2]
```diff
@@ -45,7 +45,7 @@ format:
 	npm run format:json
 
 export_food_format: export_food format
-
```
Comment (n1k0) [inline data/Makefile:48]: so it appears this file has been formatted despite `data` being blacklisted

**Review by n1k0 — APPROVED** [review]
This is a very valuable change, thanks! I tested as much scenarios as I could and couldn't spot any misbehavior.

File: data/Makefile (hunk 2 above)
Comment (vjousse) [reply data/Makefile:48]: Data is blacklisted for ruff only. Here it's the end of file formatter who removed the trailing line

File: pyproject.toml (hunk 1 above)
Comment (vjousse) [reply pyproject.toml:5]: It looks like there is a lot of code related to Jupyter notebooks and I don't want to mess up with something that could have side effects that I don't understand for now.

I suggest to keep it excluded for now until I can discuss with Paul/Christophe about those files.

Here is the ruff output for this directory FYI.

```
data/bwapi/server.py:111:5: E722 Do not use bare `except`
data/common/delete_methods.py:3:8: F401 [*] `bw2io` imported but unused
data/common/distances/distances_to_csv.py:72:41: F821 Undefined name `pycountry`
data/common/distances/distances_to_csv.py:76:18: F821 Undefined name `pycountry`
data/common/distances/distances_to_csv.py:77:22: F821 Undefined name `_`
data/common/distances/transports.py:4:27: F401 [*] `common.export.export_json` imported but unused
data/common/impacts.py:92:8: E721 Do not compare types, use `isinstance()`
data/common/import_.py:95:12: E721 Do not compare types, use `isinstance()`
data/common/import_.py:137:16: E721 Do not compare types, use `isinstance()`
data/common/import_.py:193:15: F541 [*] f-string without any placeholders
data/food/export.py:32:5: F401 [*] `food.ecosystemic_services.ecosystemic_services.plot_ecs_transformations` imported but unused
data/food/export.py:199:51: E721 Do not compare types, use `isinstance()`
data/food/export.py:201:43: E721 Do not compare types, use `isinstance()`
data/food/export.py:204:12: E721 Do not compare types, use `isinstance()`
data/import_agribalyse.py:132:5: F811 Redefinition of unused `sync_datapackages` from line 14
data/import_method.py:12:5: F401 [*] `bw2io.strategies.fix_localized_water_flows` imported but unused
data/import_method.py:15:5: F401 [*] `bw2io.strategies.migrate_exchanges` imported but unused
data/import_method.py:16:5: F401 [*] `bw2io.strategies.normalize_biosphere_categories` imported but unused
data/notebooks/explore.py:6:1: E402 Module level import not at top of file
data/notebooks/explore.py:7:1: E402 Module level import not at top of file
data/notebooks/explore.py:9:1: E402 Module level import not at top of file
data/notebooks/explore.py:12:1: E402 Module level import not at top of file
data/notebooks/explore.py:15:1: E402 Module level import not at top of file
data/notebooks/explore.py:16:1: E402 Module level import not at top of file
data/notebooks/explore.py:17:1: E402 Module level import not at top of file
data/notebooks/explore.py:18:1: E402 Module level import not at top of file
data/notebooks/explore.py:19:1: E402 Module level import not at top of file
data/notebooks/explore.py:21:1: E402 Module level import not at top of file
data/notebooks/explore.py:22:1: E402 Module level import not at top of file
data/notebooks/explore.py:23:1: E402 Module level import not at top of file
data/notebooks/explore.py:24:1: E402 Module level import not at top of file
data/notebooks/explore.py:25:1: E402 Module level import not at top of file
data/notebooks/explore.py:26:1: E402 Module level import not at top of file
data/notebooks/explore.py:27:1: E402 Module level import not at top of file
data/notebooks/explore.py:194:52: E721 Do not compare types, use `isinstance()`
data/notebooks/explore.py:202:15: E741 Ambiguous variable name: `l`
data/notebooks/explore.py:208:88: E721 Do not compare types, use `isinstance()`
data/notebooks/explore.py:373:22: F541 [*] f-string without any placeholders
data/notebooks/explore.py:448:22: F541 [*] f-string without any placeholders
data/notebooks/explore.py:491:22: F541 [*] f-string without any placeholders
data/notebooks/explore.py:601:29: F841 [*] Local variable `e` is assigned to but never used
data/notebooks/ingredients.py:6:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:7:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:11:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:14:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:15:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:16:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:17:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:19:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:20:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:21:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:22:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:23:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:24:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:25:1: E402 Module level import not at top of file
data/notebooks/ingredients.py:472:5: F821 Undefined name `display`
data/notebooks/ingredients.py:488:9: F821 Undefined name `display`
data/notebooks/ingredients.py:665:9: F821 Undefined name `display`
data/notebooks/ingredients.py:674:9: F821 Undefined name `display`
data/notebooks/ingredients.py:705:9: F821 Undefined name `display`
data/notebooks/ingredients.py:711:9: F821 Undefined name `display`
data/notebooks/ingredients.py:722:9: F821 Undefined name `display`
data/notebooks/ingredients.py:733:9: F821 Undefined name `display`
data/notebooks/ingredients.py:744:9: F821 Undefined name `display`
data/notebooks/ingredients.py:753:9: F821 Undefined name `display`
data/notebooks/ingredients.py:759:9: F821 Undefined name `display`
data/notebooks/ingredients.py:769:5: F821 Undefined name `display`
data/notebooks/ingredients.py:801:5: F821 Undefined name `display`
data/notebooks/ingredients.py:812:9: F821 Undefined name `display`
data/notebooks/ingredients.py:820:9: F821 Undefined name `display`
data/notebooks/ingredients.py:827:9: F821 Undefined name `display`
data/notebooks/ingredients.py:856:9: F821 Undefined name `display`
data/notebooks/ingredients.py:859:5: F821 Undefined name `display`
data/notebooks/ingredients.py:867:9: F821 Undefined name `display`
data/notebooks/ingredients.py:876:9: F821 Undefined name `display`
data/notebooks/ingredients.py:893:9: F821 Undefined name `display`
data/notebooks/ingredients.py:903:9: F821 Undefined name `display`
data/notebooks/ingredients.py:913:9: F821 Undefined name `display`
data/notebooks/ingredients.py:918:9: F821 Undefined name `display`
data/notebooks/ingredients.py:938:1: F821 Undefined name `display`
data/notebooks/ingredients.py:1268:34: F541 [*] f-string without any placeholders
data/notebooks/textile_processes.py:74:1: F821 Undefined name `display`
data/spapi/simapro.py:59:13: E722 Do not use bare `except`
data/textile/activities.py:75:36: E721 Do not compare types, use `isinstance()`
data/textile/export.py:32:12: E721 Do not compare types, use `isinstance()`
```

File: data/bwapi/README.txt
Comment (vjousse) [reply data/bwapi/README.txt:1]: Same, this is not ruff, it's the line ending checker

**Pull request merged**